compiler/rustc_passes/src/liveness.rs

```diff
@@ -329,6 +329,44 @@ impl<'tcx> Visitor<'tcx> for IrMaps<'tcx> {
             }
         }
 
+        // Allow todo! macro
+        /*
+            Skips checking for unused variables when the trailing expression
+            of the body is core::panicking::panic("not yet implemented"), ie
+            as emitted by todo!().
+
+            # Example
+
+            fn foo(x: i32) {
+                // arbitrary code
+                todo!()
+            }
+        */
+        if let ExprKind::Block(block, _) = &body.value.kind {
+            if let Some(expr) = block.expr {
+                if let ExprKind::Call(call, [arg]) = expr.kind {
+                    if let ExprKind::Path(QPath::Resolved(_, path)) = call.kind {
+                        if matches!(&path.res, Res::Def(DefKind::Fn, _)) {
+                            // FIXME: Better way of extracting literal
+                            let debug = format!("{:?}", &path.res);
+                            if (debug.contains("std[") || debug.contains("core["))
+                                && debug.contains("]::panicking::panic)")
+                            {
+                                if let ExprKind::Lit(spanned) = &arg.kind {
+                                    if let LitKind::Str(symbol, StrStyle::Cooked) = spanned.node {
+                                        // FIXME: Better way of matching symbol
+                                        if &*symbol.as_str() == "not yet implemented" {
+                                            return;
+                                        }
+                                    }
+                                }
+                            }
+                        }
```

Comment:
This should be a less hacky way of determining if it calls panic:

```suggestion
                        if matches!(&path.res, Res::Def(DefKind::Fn, path_def_id)) {
                            let begin_panic_def_id = self.tcx.lang_items().begin_panic_fn();
                            if begin_panic_def_id == Some(path_def_id) {
                                // FIXME: Better way of extracting literal
                                if let ExprKind::Lit(spanned) = &arg.kind {
                                    if let LitKind::Str(symbol, StrStyle::Cooked) = spanned.node {
                                        // FIXME: Better way of matching symbol
                                        if &*symbol.as_str() == "not yet implemented" {
                                            return;
                                        }
                                    }
                                }
                            }
                        }
```

Reply:
However, note that neither the current code nor the code I suggested will catch invocations like `todo!("some message")` that override the panic message.